File: R/q1_io.R
# R/q1_io.R

suppressWarnings({
  library(readxl)
  library(dplyr)
})

require_columns <- function(df, cols, where = "") {
  missing <- setdiff(cols, names(df))
  if (length(missing) > 0) {
    stop(
      sprintf(
        "Missing columns in %s: %s. Available: %s",
        where,
        paste(missing, collapse = ", "),
        paste(names(df), collapse = ", ")
      ),
      call. = FALSE
    )
  }
  invisible(df)
}

read_x_strict_one_sheet <- function(path) {
  sheets <- readxl::excel_sheets(path)
  if (length(sheets) != 1) {
    stop(sprintf("x.xlsx must have exactly 1 sheet, found: %s", paste(sheets, collapse = ", ")))
  }
  readxl::read_excel(path, sheet = sheets[1]) %>%
    mutate(date = as.Date(date))
}

read_y_data_y_sheet <- function(path, sheet = "data_y") {
  sheets <- readxl::excel_sheets(path)
  if (!(sheet %in% sheets)) {
    stop(sprintf("y.xlsx missing sheet '%s'. Found: %s", sheet, paste(sheets, collapse = ", ")))
  }
  readxl::read_excel(path, sheet = sheet) %>%
    mutate(date = as.Date(date))
}

read_descriptions_sheet <- function(path, sheet = "descriptions") {
  sheets <- readxl::excel_sheets(path)
  if (!(sheet %in% sheets)) {
    return(NULL)  # not an error, but then DESCRIPTION will be NA
  }
  readxl::read_excel(path, sheet = sheet) %>%
    rename(CODE = 1, DESCRIPTION = 2)
}

write_csv_strict <- function(df, path) {
  dir.create(dirname(path), recursive = TRUE, showWarnings = FALSE)
  write.csv(df, path, row.names = FALSE)
}

File: python/q1_io.py
# python/q1_io.py
from __future__ import annotations

from pathlib import Path
import pandas as pd


def read_x_strict_one_sheet(path: Path) -> pd.DataFrame:
    xls = pd.ExcelFile(path)
    if len(xls.sheet_names) != 1:
        raise ValueError(
            f"{path.name} must have exactly 1 sheet, but got {len(xls.sheet_names)}: {xls.sheet_names}"
        )
    return xls.parse(xls.sheet_names[0])


def read_y_sheet_strict(path: Path, sheet_name: str = "data_y") -> pd.DataFrame:
    xls = pd.ExcelFile(path)
    if sheet_name not in xls.sheet_names:
        raise ValueError(f"{path.name} must contain sheet '{sheet_name}'. Available: {xls.sheet_names}")
    return xls.parse(sheet_name)


def read_desc_sheet_strict(path: Path, sheet_name: str = "descriptions") -> pd.DataFrame:
    xls = pd.ExcelFile(path)
    if sheet_name not in xls.sheet_names:
        raise ValueError(f"{path.name} must contain sheet '{sheet_name}'. Available: {xls.sheet_names}")
    df = xls.parse(sheet_name)
    if not {"CODE", "DESCRIPTION"}.issubset(df.columns):
        raise ValueError(f"Sheet '{sheet_name}' must contain CODE and DESCRIPTION. Got: {list(df.columns)}")
    return df[["CODE", "DESCRIPTION"]]


def require_columns(df: pd.DataFrame, required: list[str], context: str) -> None:
    missing = [c for c in required if c not in df.columns]
    if missing:
        raise ValueError(f"Missing columns in {context}: {missing}. Available: {list(df.columns)}")


def write_csv(df: pd.DataFrame, out_path: Path) -> None:
    out_path.parent.mkdir(parents=True, exist_ok=True)
    df.to_csv(out_path, index=False)
